flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2116,8 +2171,34 @@ struct StoreOpConversion : public FIROpConversion<fir::StoreOp> {
   mlir::LogicalResult
   matchAndRewrite(fir::StoreOp store, OperandTy operands,
                   mlir::ConversionPatternRewriter &rewriter) const override {
-    rewriter.replaceOpWithNewOp<mlir::LLVM::StoreOp>(store, operands[0],
-                                                     operands[1]);
+    if (store.value().getType().isa<fir::BoxType>()) {
+      // The box value is in memory, load it first before storing it into
+      // the destination address.
+      auto loc = store.getLoc();
+      auto voidTy = getVoidPtrType(store.getContext());
+      auto src =
+          rewriter.create<mlir::LLVM::BitcastOp>(loc, voidTy, operands[0]);
+      auto dst =
+          rewriter.create<mlir::LLVM::BitcastOp>(loc, voidTy, operands[1]);
+      auto falseAttr = rewriter.getBoolAttr(false);
+      auto i1Ty = mlir::LLVM::LLVMType::getInt1Ty(store.getContext());
+      auto isVolatile =
+          rewriter.create<mlir::LLVM::ConstantOp>(loc, i1Ty, falseAttr);
+      auto i64Ty = mlir::LLVM::LLVMType::getInt64Ty(store.getContext());
+      // Get the length of descriptor type by using getelementPtr. This assumes
+      // The descriptor type provided to LLVM is accurate (e.g if there is an
+      // addendum, it must be visible in llvm type).
+      auto boxPtrTy = unwrap(operands[0].getType());
+      auto null = rewriter.create<mlir::LLVM::NullOp>(loc, boxPtrTy);
+      auto one = genConstantIndex(loc, i64Ty, rewriter, 1);
+      auto sizePtr = genGEP(loc, boxPtrTy, rewriter, null, one);
+      auto len = rewriter.create<mlir::LLVM::PtrToIntOp>(loc, i64Ty, sizePtr);
+      rewriter.replaceOpWithNewOp<mlir::LLVM::MemcpyOp>(store, dst, src, len,
```

Comment:
Have you tried making box values be first-class LLVM struct (register) values? It would mean some rework on lowering to LLVM IR, of course, but it might reduce some of this special-case code.

The one case that has to change off the top of my head is the automatic conversion of box values to referenceable objects (in memory). `embox` would be trivially without effects, which would allow CSE and code motion, which would be nice. The cost would be converting values of box types to memory references at call sites.

Reply:
I am happy to explore that alternative. First, I guess I need to think of a few fir programs doing embox and set some optimization expectation to compare the alternatives.

Reply:
I pursuing the investigation in parallel. My first observations seems to indicate that we would have more room to optimize stack storage by making temporary box allocation explicit in fir (as opposed as hidden in the emboxOp). However, it does not seems to impact timing significantly to do one or the other.